Userland/Libraries/LibWeb/Fetch/FetchMethod.cpp

```diff
@@ -0,0 +1,179 @@
+    //     substeps:
+    auto process_response = [&vm, locally_aborted, promise_capability, request, response_object_handle, &relevant_realm](NonnullRefPtr<Infrastructure::Response> response) mutable {
+        // 1. If locallyAborted is true, terminate these substeps.
+        if (locally_aborted->value())
+            return;
+
+        // 2. If response’s aborted flag is set, then:
+        if (response->aborted()) {
+            // FIXME: 1. Let deserializedError be the result of deserialize a serialized abort reason given controller’s
+            //           serialized abort reason and relevantRealm.
+            auto deserialized_error = JS::js_undefined();
+
+            // 2. Abort the fetch() call with p, request, responseObject, and deserializedError.
+            abort_fetch(vm, promise_capability, request, response_object_handle.cell(), deserialized_error);
+            return;
+        }
+
+        // 3. If response is a network error, then reject p with a TypeError and terminate these substeps.
+        if (response->is_network_error()) {
+            auto message = response->network_error_message().value_or("Response is a network error"sv);
+            WebIDL::reject_promise(vm, promise_capability, JS::TypeError::create(relevant_realm, message));
+            return;
+        }
+
+        // 4. Set responseObject to the result of creating a Response object, given response, "immutable", and
+        //    relevantRealm.
+        auto response_object = Response::create(move(response), Headers::Guard::Immutable, relevant_realm);
+        response_object_handle = JS::make_handle(response_object.ptr());
+
+        // 5. Resolve p with responseObject.
+        WebIDL::resolve_promise(vm, promise_capability, response_object);
+    };
+    controller = MUST(Fetching::fetch(
+        realm,
+        request,
+        Infrastructure::FetchAlgorithms::create({
+            .process_request_body_chunk_length = {},
+            .process_request_end_of_body = {},
+            .process_early_hints_response = {},
+            .process_response = move(process_response),
+            .process_response_end_of_body = {},
+            .process_response_consume_body = {},
+        })));
+
+    // 13. Return p.
+    return verify_cast<JS::Promise>(*promise_capability->promise().ptr());
+}
+
+// https://fetch.spec.whatwg.org/#abort-fetch
+void abort_fetch(JS::VM& vm, JS::PromiseCapability const& promise_capability, NonnullRefPtr<Infrastructure::Request> request, JS::GCPtr<Response> response_object, JS::Value error)
+{
+    dbgln_if(WEB_FETCH_DEBUG, "Fetch: Aborting fetch with: request @ {}, error = {}", request.ptr(), error);
+
+    // 1. Reject promise with error.
+    // NOTE: This is a no-op if promise has already fulfilled.
+    WebIDL::reject_promise(vm, promise_capability, error);
+
+    // 2. If request’s body is not null and is readable, then cancel request’s body with error.
+    if (request->body().has<Infrastructure::Body>()) {
```

Comment:
Can only `Body` be readable?
Normally you check this with `!...has<Empty>())`
Because I think it can be bytebuffer?

Also if not can use if with init

Reply:
> Because I think it can be bytebuffer?

from https://fetch.spec.whatwg.org/#concept-request-body:

> A [byte sequence](https://infra.spec.whatwg.org/#byte-sequence) will be [safely extracted](https://fetch.spec.whatwg.org/#bodyinit-safely-extract) into a [body](https://fetch.spec.whatwg.org/#concept-body) early on in [fetch](https://fetch.spec.whatwg.org/#concept-fetch). As part of [HTTP fetch](https://fetch.spec.whatwg.org/#concept-http-fetch) it is possible for this field to be set to null due to certain redirects.

> Also if not can use if with init

Will do that though :)